In [1]:
pip install lxml

Note: you may need to restart the kernel to use updated packages.


In [43]:
import pandas as pd
import numpy as np
import requests
import json
from pandas.io.json import json_normalize

In [3]:
!wget -q -O 'cdata.html' https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
print('Data downloaded!')

Data downloaded!


#### Getting relevant data from website.

In [4]:
df= pd.read_html('cdata.html')
y = df[0]
print(y)

    Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       

# Explore 

#### Changing column name Postal Code  to Postalcode

In [5]:
#The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
y.rename(columns = {'Postal Code': 'Postalcode'}, inplace = True)
y.head()

Postalcode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

In [6]:
#Checking records with NaN/Null values for Neighborhood column
y['Neighborhood'].isnull().sum()

0

#### Ignore cells of borough that is "Not assigned"

In [7]:
# Only process the cells that have an assigned borough and list the number of rows. 
cdf = y[y['Borough'] != 'Not assigned']
cdf.shape

(103, 3)

In [8]:
cdf.head()

Postalcode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

#### Dataset is already combined with Neighborhoods with similar Postalcode. No recurring records with same postal code. Cross verified by getting "no of rows" before running "unique" method on column and after (103 before and after)

In [9]:
#More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

cdf['Postalcode'].unique().shape

(103,)

#### There are no neighborhood's with null values 

In [10]:
#If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
cdf['Neighborhood'].isnull().sum()

0

#### Number of rows and columns in the dataframe

In [11]:
cdf.shape

(103, 3)

# Segment

In [14]:
#getting data for latitude & longitude from the csv file
path = 'http://cocl.us/Geospatial_data'
ldf = pd.read_csv(path)
ldf.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

#### Merging dataframes to get relevant latitude longitude for the Postal codes

In [18]:
ndf = pd.merge(cdf, ldf, how='left', left_on=['Postalcode'], right_on=['Postal Code'])
ndf.head()

Postalcode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

  Postal Code   Latitude  Longitude  
0         M3A  43.753259 -79.329656  
1         M4A  43.725882 -79.315572  
2         M5A  43.654260 -79.360636  
3         M6A  43.718518 -79.464763  
4         M7A  43.662301 -79.389494

#### Getting boroughs that contain the word "Toronto"

In [34]:
toronto_data = ndf[ndf['Borough'].str.contains("Toronto")].reset_index(drop = True)
toronto_data.head()

Postalcode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

  Postal Code   Latitude  Longitude  
0         M5A  43.654260 -79.360636  
1         M7A  43.662301 -79.389494  
2         M5B  43.657162 -79.378937  
3         M5C  43.651494 -79.375418  
4         M4E  43.676357 -79.293031

In [35]:
toronto_data.shape

(39, 6)

In [24]:
pip install geopy

     |████████████████████████████████| 122kB 3.5MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


#### Use geopy library to get the latitude and longitude values of Toronto City.

In [27]:
from geopy.geocoders import Nominatim
address = 'Toronto, TO'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


### Importing relevant modules

In [29]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

#### Create a map of Toronto with data superimposed on top

In [36]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Define Foursquare Credentials and Version

In [32]:
CLIENT_ID = 'GKDTMTKVUR2AKLHRS0YXZ4KB2AKROAZLA5TSAGRISI0GTN2V' # your Foursquare ID
CLIENT_SECRET = 'ZV0EWRCCUZRJU1WHJT10E0SACFYTF1KFPU2BBTBDVVQESRWW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GKDTMTKVUR2AKLHRS0YXZ4KB2AKROAZLA5TSAGRISI0GTN2V
CLIENT_SECRET:ZV0EWRCCUZRJU1WHJT10E0SACFYTF1KFPU2BBTBDVVQESRWW


#### Getting first neighborhoods in our dataframe

In [37]:
toronto_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

#### Latitude and Longitude of first neighborhoods

In [38]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


#### Now, let's get the top 100 venues that are in Regent Park within a radius of 500 meters.

In [39]:
#  let's create the GET request URL. Name your URL url
search_query = 'Regent Park'
radius = 500
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=GKDTMTKVUR2AKLHRS0YXZ4KB2AKROAZLA5TSAGRISI0GTN2V&client_secret=ZV0EWRCCUZRJU1WHJT10E0SACFYTF1KFPU2BBTBDVVQESRWW&ll=43.6534817,-79.3839347&v=20180605&query=Regent Park&radius=500&limit=100'

In [44]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ed75af2a2e538001b16115b'},
 'response': {'venues': [{'id': '535019a4498ea2ba59289985',
    'name': 'Standard Parking',
    'location': {'address': '181 Bay Street, P1 Level',
     'crossStreet': 'Enter on 10 Front St W or 7 Wellington St W.',
     'lat': 43.653226,
     'lng': -79.383184,
     'labeledLatLngs': [{'label': 'display',
       'lat': 43.653226,
       'lng': -79.383184}],
     'distance': 66,
     'postalCode': 'M5J 2T3',
     'cc': 'CA',
     'city': 'Toronto',
     'state': 'ON',
     'country': 'Canada',
     'formattedAddress': ['181 Bay Street, P1 Level (Enter on 10 Front St W or 7 Wellington St W.)',
      'Toronto ON M5J 2T3',
      'Canada']},
    'categories': [{'id': '4c38df4de52ce0d596b336e1',
      'name': 'Parking',
      'pluralName': 'Parking',
      'shortName': 'Parking',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/parking_',
       'suffix': '.png'},
      'primary': True}],
    'referral

#### Normalizing the json values

In [46]:
venues = results['response']['venues']
nearby_venues = json_normalize(venues) # flatten JSON
nearby_venues.head()

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  


id                                name  \
0  535019a4498ea2ba59289985                    Standard Parking   
1  53b1a1e4498e241b0ac7b5e5     City Of Toronto Parking Permits   
2  51803df8e4b0f7cc47bb1531                   Larry Sefton Park   
3  4b9177e4f964a520d2be33e3  TTC Streetcar #403 - Victoria park   
4  557f24f6498e5325baa2a8c4          Trinity Park And Labyrinth   

                                          categories    referralId  hasPerk  \
0  [{'id': '4c38df4de52ce0d596b336e1', 'name': 'P...  v-1591172098    False   
1  [{'id': '4bf58dd8d48988d129941735', 'name': 'C...  v-1591172098    False   
2  [{'id': '4bf58dd8d48988d164941735', 'name': 'P...  v-1591172098    False   
3  [{'id': '4bf58dd8d48988d1fc931735', 'name': 'L...  v-1591172098    False   
4  [{'id': '4bf58dd8d48988d15a941735', 'name': 'G...  v-1591172098    False   

           location.address                          location.crossStreet  \
0  181 Bay Street, P1 Level  Enter on 10 Front St W or 7 Wellington St W.   
1     100 Queen Street West                                           Bay   
2                       NaN                                           NaN   
3                       NaN                                           NaN   
4                   Toronto                                        Kanada   

   location.lat  location.lng  \
0     43.653226    -79.383184   
1     43.653912    -79.384590   
2     43.654589    -79.383405   
3     43.650735    -79.386921   
4     43.654224    -79.381882   

                             location.labeledLatLngs  location.distance  \
0  [{'label': 'display', 'lat': 43.653226, 'lng':...                 66   
1  [{'label': 'display', 'lat': 43.65391159057617...                 71   
2  [{'label': 'display', 'lat': 43.65458899984554...                130   
3  [{'label': 'display', 'lat': 43.650735, 'lng':...                389   
4  [{'label': 'display', 'lat': 43.65422439575195...                184   

  location.postalCode location.cc location.city location.state  \
0             M5J 2T3          CA       Toronto             ON   
1                 NaN          CA       Toronto             ON   
2                 NaN          CA           NaN            NaN   
3                 NaN          CA       Toronto             ON   
4                 NaN          CA       Toronto             ON   

  location.country                          location.formattedAddress  \
0           Canada  [181 Bay Street, P1 Level (Enter on 10 Front S...   
1           Canada  [100 Queen Street West (Bay), Toronto ON, Canada]   
2           Canada                                           [Canada]   
3           Canada                               [Toronto ON, Canada]   
4           Canada             [Toronto (Kanada), Toronto ON, Canada]   

  venuePage.id location.neighborhood  
0          NaN                   NaN  
1          NaN                   NaN  
2          NaN                   NaN  
3          NaN                   NaN  
4          NaN                   NaN

#### function that extracts the category of the venue

In [47]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

#### clean the json and structure it into a pandas dataframe

In [48]:
# filter columns
filtered_columns = ['name', 'categories', 'location.lat', 'location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name          categories        lat  \
0                    Standard Parking             Parking  43.653226   
1     City Of Toronto Parking Permits           City Hall  43.653912   
2                   Larry Sefton Park               Plaza  43.654589   
3  TTC Streetcar #403 - Victoria park  Light Rail Station  43.650735   
4          Trinity Park And Labyrinth              Garden  43.654224   

         lng  
0 -79.383184  
1 -79.384590  
2 -79.383405  
3 -79.386921  
4 -79.381882

In [49]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

47 venues were returned by Foursquare.


#### Function to repeat the same process to all the neighborhoods in Toronto

In [50]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### The above function is run on each neighborhood and created a new dataframe called *toronto_venues*.

In [52]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [53]:
print(toronto_venues.shape)
toronto_venues.head()

(1616, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

##### No of venues returned for each neighborhood

In [54]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     66   
Christie                                                               17   
Church and Wellesley                                                   77   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           16   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         22   
Kensington Market, Chinatown, Grange Park                              58   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             3   
North Toronto West, Lawrence Park                                      20   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     37   
St. James Town                                                         80   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         94   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
Brockton, Parkdale Village, Exhibition Place                            22   
Business reply mail Processing Centre, South Ce...                      17   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      66   
Christie                                                                17   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                    

### Analyze each neighborhood

In [55]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Toy / Game Store  Trail  Train Station  \
0             0         0  ...                 0      0              0   
1             0         0  ...                 0      0              0   
2             0         0  ...                 0      0              0   
3             0         0  ...                 0      0              0   
4             0         0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Wine Bar  Wings Joint  Women's Store  
0                      0         0            0              0  
1                      0         0            0              0  
2                      0         0            0              0  
3                      0         0            0              0  
4                      0         0            0              0  

[5 rows x 235 columns]

In [56]:
toronto_onehot.shape

(1616, 235)

#### Grouping rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [57]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.058824   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.015152   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.025974   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.022727   
20                        Moore Park, Summerhill East     0.000000   
21                  North Toronto West, Lawrence Park     0.050000   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.030303   
24                          Regent Park, Harbourfront     0.022222   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.027027   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010638   
32                                    Studio District     0.024390   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.023810   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.027778   

    Afghan Restaurant   Airport  Airport Food Court  Airport Lounge  \
0            0.000000  0.000000            0.000000        0.000000   
1            0.000000  0.000000            0.000000        0.000000   
2            0.000000  0.000000            0.000000        0.000000   
3            0.000000  0.066667            0.066667        0.133333   
4            0.000000  0.000000            0.000000        0.000000   
5            0.000000  0.000000            0.000000        0.000000   
6            0.012987  0.000000            0.000000        0.000000   
7            0.000000  0.000000            0.000000        0.000000   
8            0.000000  0.000000            0.000000        0.000000   
9            0.000000  0.000000            0.000000        0.000000   
10           0.000000  0.000000            0.000000        0.000000   
11           0.000000  0.000000            0.000000        0.000000   
12           0.000000  0.000000            0.000000        0.000000   
13           0.000000  0.000000            0.000000        0.000000   
14           0.000000  0.000000            0.000000        0.000000   
15           0.000000  0.000000            0.000000        0.000000   
16           0.000000  0.000000  

In [59]:
toronto_grouped.shape

(39, 235)

#### Let's print each neighborhood along with the top 5 most common venues

In [60]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2                Café  0.04
3         Cheese Shop  0.04
4  Seafood Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
                venue  freq
0                Café  0.14
1         Coffee Shop  0.09
2      Breakfast Spot  0.09
3       Grocery Store  0.05
4  Italian Restaurant  0.05


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0  Light Rail Station  0.12
1         Yoga Studio  0.06
2       Auto Workshop  0.06
3                Park  0.06
4         Pizza Place  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.13
1   Airport Service  0.13
2  Airport Terminal  0.13
3          Boutique  0.07
4     Boat or Ferry  0.07


----Central Bay Street----
                 venue  

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order

In [63]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [65]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...      Airport Terminal   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0          Cocktail Bar            Restaurant            Cheese Shop   
1        Breakfast Spot           Coffee Shop  Performing Arts Venue   
2           Yoga Studio            Comic Shop          Burrito Place   
3        Airport Lounge       Airport Service          Boat or Ferry   
4                  Café    Italian Restaurant         Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar                   Pub                Bakery   
1          Climbing Gym                   Bar          Intersection   
2            Skate Park  Fast Food Restaurant         Auto Workshop   
3           Coffee Shop      Sculpture Garden                   Bar   
4   Japanese Restaurant            Restaurant           Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café    Seafood Restaurant         Breakfast Spot  
1                Bakery               Stadium     Italian Restaurant  
2        Farmers Market           Pizza Place                Brewery  
3       Harbor / Marina    Airport Food Court                Airport  
4      Department Store          Burger Joint        Bubble Tea Shop

## Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [66]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [67]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postalcode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

  Postal Code   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0         M5A  43.654260 -79.360636               0           Coffee Shop   
1         M7A  43.662301 -79.389494               0           Coffee Shop   
2         M5B  43.657162 -79.378937               0        Clothing Store   
3         M5C  43.651494 -79.375418               0                  Café   
4         M4E  43.676357 -79.293031               0                 Trail   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0                  Park                     Bakery                   Pub   
1      Sushi Restaurant                Yoga Studio     College Cafeteria   
2           Coffee Shop  Middle Eastern Restaurant       Bubble Tea Shop   
3           Coffee Shop                 Restaurant          Cocktail Bar   
4                   Pub          Health Food Store   Distribution Center   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot            Restaurant                  Café   
1              Beer Bar         Smoothie Shop        Sandwich Place   
2   Japanese Restaurant    Italian Restaurant        Cosmetics Shop   
3   American Restaurant             Gastropub        Cosmetics Shop   
4         Deli / Bodega      Department Store          Dessert Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Theater    Mexican Restaurant             Shoe Store  
1          Burger Joint         Burrito Place                   Café  
2                  Café        Lingerie Store                  Hotel  
3    Italian Restaurant      Department Store         Clothing Store  
4                 Diner        Discount Store          Women's Store

### visualize the resulting clusters

In [68]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine CLusters

#### Cluster #1

In [69]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Longitude  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto -79.360636               0           Coffee Shop   
1   Downtown Toronto -79.389494               0           Coffee Shop   
2   Downtown Toronto -79.378937               0        Clothing Store   
3   Downtown Toronto -79.375418               0                  Café   
4       East Toronto -79.293031               0                 Trail   
5   Downtown Toronto -79.373306               0           Coffee Shop   
6   Downtown Toronto -79.387383               0           Coffee Shop   
7   Downtown Toronto -79.422564               0         Grocery Store   
8   Downtown Toronto -79.384568               0           Coffee Shop   
9       West Toronto -79.442259               0                Bakery   
10  Downtown Toronto -79.381752               0           Coffee Shop   
11      West Toronto -79.419750               0                   Bar   
12      East Toronto -79.352188               0      Greek Restaurant   
13  Downtown Toronto -79.381576               0           Coffee Shop   
14      West Toronto -79.428191               0                  Café   
15      East Toronto -79.315572               0        Sandwich Place   
16  Downtown Toronto -79.379817               0           Coffee Shop   
17      East Toronto -79.340923               0                  Café   
20   Central Toronto -79.390197               0      Department Store   
22      West Toronto -79.464763               0       Thai Restaurant   
23   Central Toronto -79.405678               0        Clothing Store   
24   Central Toronto -79.405678               0                  Café   
25      West Toronto -79.456325               0             Gift Shop   
26   Central Toronto -79.388790               0          Dessert Shop   
27  Downtown Toronto -79.400049               0                  Café   
28      West Toronto -79.484450               0                  Café   
30  Downtown Toronto -79.400049               0                  Café   
31   Central Toronto -79.400049               0                   Pub   
32  Downtown Toronto -79.394420               0      Airport Terminal   
34  Downtown Toronto -79.374846               0           Coffee Shop   
35  Downtown Toronto -79.367675               0           Coffee Shop   
36  Downtown Toronto -79.382280               0           Coffee Shop   
37  Downtown Toronto -79.383160               0           Coffee Shop   
38      East Toronto -79.321558               0    Light Rail Station   

   2nd Most Common Venue      3rd Most Common Venue  \
0                   Park                     Bakery   
1       Sushi Restaurant                Yoga Studio   
2            Coffee Shop  Middle Eastern Restaurant   
3            Coffee Shop                 Restaurant   
4                    Pub          Health Food Store   
5           Cocktail Bar                 Restaurant   
6                   Café         Italian Restaurant   
7                   Café                       Park   
8                   Café                 Restaurant   
9               Pharmacy                Coffee Shop   
10              Aquarium                      Hotel   
11      Asian Restaurant                       Café   
12           Coffee Shop         Italian Restaurant   
13                  Café                      Hotel   
14        Breakfast Spot                Coffee Shop   
15  Fast Food Restaurant                       Park   
16                  Café                 Restaurant   
17           Coffee Shop                  Gastropub   
20                  Park                      Hotel   
22    Mexican Restaurant                       Café   
23           Coffee Shop                  Pet Store   
24        Sandwich Place                Coffee Shop   
25        Breakfast Spot              Movie Theater   
26        Sandwich Place                       Café   
27            Restaurant                        Bar   
28           Coffee Shop   

#### Cluster #2

In [70]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Longitude  Cluster Labels 1st Most Common Venue  \
21   Central Toronto -79.411307               1                  Park   
33  Downtown Toronto -79.377529               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21         Jewelry Store                 Trail      Sushi Restaurant   
33            Playground                 Trail      Cuban Restaurant   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21                Women's Store         Deli / Bodega     Electronics Store   
33  Eastern European Restaurant            Donut Shop      Doner Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21  Eastern European Restaurant            Donut Shop       Doner Restaurant  
33                      Dog Run   Distribution Center         Discount Store

#### Cluster #3

In [73]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Longitude  Cluster Labels 1st Most Common Venue  \
18  Central Toronto  -79.38879               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18           Swim School              Bus Line           Event Space   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
18     Electronics Store  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18      Doner Restaurant               Dog Run    Distribution Center

#### Cluster #4

In [74]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Longitude  Cluster Labels 1st Most Common Venue  \
19  Central Toronto -79.416936               3                  Pool   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden         Women's Store          Cupcake Shop   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
19     Electronics Store  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19      Doner Restaurant               Dog Run    Distribution Center

#### Cluster #5

In [78]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Longitude  Cluster Labels 1st Most Common Venue  \
29  Central Toronto  -79.38316               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29           Summer Camp            Restaurant          Cupcake Shop   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
29  Eastern European Restaurant            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29               Dog Run   Distribution Center         Discount Store